### PR TITLE
Add identifiers to virtio devices

### PR DESCRIPTION
**Pull request by sboeuf**
As a preparatory work to support a full tree containing the resources associated with each device, which is going to be mandatory to perform proper snapshot/restore, we need to identify each virtio device with a unique id.

### CODE REVIEW

File: src/main.rs [hunk 1]
```diff
@@ -147,7 +147,8 @@ fn create_app<'a, 'b>(
             Arg::with_name("rng")
                 .long("rng")
                 .help(
-                    "Random number generator parameters \"src=<entropy_source_path>,iommu=on|off\"",
+                    "Random number generator parameters \
+                    \"src=<entropy_source_path>,iommu=on|off,id=<device_id>\"",
```
Comment (sameo) [inline src/main.rs:151]: I think we should present the user with the option of defining a device id only when it makes sense, i.e. when the user can hot-plug and unplug it. Since we already added that for our device hotplug implementation, I believe our UI should not be modified with this PR.

The sole reason why we need a device ID for e.g. the RNG device here is for our internal restoring implementation and thus this should be kept internal. Same for our vsocks devices.

That means there's a part of our VmConfig that we're not exposing to the user and I believe it is just fine. In case there's a disagreement about that and we want all our config items to be visible and modifiable from the UI, then we need to have some sort of config overlay (let's call it `VmDefinition` for the sake of that discussion).for adding our internal data to the user visible configuration. And we'd snapshot that overlay and restore from it, not from the VM config. This would clearly and cleanly separate the VM config defined and modified by the user from what we internally add to it, but otoh would add some complexity as e.g. we'd need to create and restore from the `VmDefinition` instead of what we do today from `VmConfig`. Since `VmDefinition` would be an overlay on top of `VmConfig`, it would be fairly easy to build the former from the latter and amend it as e.g. we add devices that don't have a user-visible/config-exposed id.

Bottom line: For now I'm personally ok if we overload our `VmConfig` with user invisible data, but we can think about a `VmConfig` overlay for a cleaner implementation in the future.

File: vmm/src/device_manager.rs [hunk 2]
```diff
@@ -564,6 +584,9 @@ pub struct DeviceManager {
     // Backends that have been spawned
     vhost_user_backends: Vec<ActivatedBackend>,
 
+    // Counter to keep track of the consumed device IDs.
+    device_id_cnt: Wrapping<usize>,
+
```
Comment (sameo) [inline vmm/src/device_manager.rs:589]: Nit: Any reason why this is moved up?

Comment (sboeuf) [reply vmm/src/device_manager.rs:589]: Oh that's because it's not PCI specific anymore, so I thought moving it up would make this clearer. But I can move it back if you want.

File: src/main.rs (hunk 1 above)
Comment (sboeuf) [reply src/main.rs:151]: The point of having an `id` field in the `VmConfig` is to rely on it to find out if there's a corresponding entry in the device tree or not, which means we can restore from it, or simply creating the device from scratch.
The only devices for which we can accept no `id` field in the `VmConfig` are the single devices like `rng`, `console` and `mem`.
Since we can create multiple `vsock` devices, we need to store the information in the `VmConfig` so that we can do this comparison with the device tree during the restore path.

The overlay layer might be an option by making it part of the `DeviceManagerState`. That would store an extended version of the `VmConfig`, but this is gonna grow our complexity.

About the unplug, it's taken care of as part of this PR, as I find the virtio device type associated with the id, and I return an error if the type is not expected to be removed.

Comment (sameo) [reply src/main.rs:151]: > The point of having an `id` field in the `VmConfig` is to rely on it to find out if there's a corresponding entry in the device tree or not, which means we can restore from it, or simply creating the device from scratch.

I'm certainly not arguing against the idea of having each device uniquely identified to be able to restore from it.

> The only devices for which we can accept no `id` field in the `VmConfig` are the single devices like `rng`, `console` and `mem`.

Could we remove them for those devices?

> Since we can create multiple `vsock` devices, we need to store the information in the `VmConfig` so that we can do this comparison with the device tree during the restore path.
>
>
> The overlay layer might be an option by making it part of the `DeviceManagerState`. That would store an extended version of the `VmConfig`, but this is gonna grow our complexity.
>

Then we should at least not allow the user to set the `vsock` device id.
 
> About the unplug, it's taken care of as part of this PR, as I find the virtio device type associated with the id, and I return an error if the type is not expected to be removed.

Yes, and that's fine for now. But the fact that we can potentially get a request to unplug a vsock device means we could improve this design in the future, by keeping ids for static devices completely unreachable to the user.

Comment (rbradford) [reply src/main.rs:151]: I don't think we should let the user specify an id in this case. For "internal" ids we should prefix them with an `_` and never let the user specify anything with an `_` so that we know the use can't specify one that would clash with our "internal" ones.

File: src/main.rs [hunk 3]
```diff
@@ -196,7 +197,8 @@ fn create_app<'a, 'b>(
             Arg::with_name("vsock")
                 .long("vsock")
                 .help(
-                    "Virtio VSOCK parameters \"cid=<context_id>,sock=<socket_path>,iommu=on|off\"",
+                    "Virtio VSOCK parameters \
+                    \"cid=<context_id>,sock=<socket_path>,iommu=on|off,id=<device_id>\"",
```
Comment (rbradford) [inline src/main.rs:201]: I don't think we should expose this to the user here as I don't think the user can use that.

File: src/main.rs [hunk 4]
```diff
@@ -516,6 +518,7 @@ mod unit_tests {
                 rng: RngConfig {
                     src: PathBuf::from("/dev/urandom"),
                     iommu: false,
+                    id: None,
```
Comment (rbradford) [inline src/main.rs:521]: For things like rng where is is static i.e. `_rng0` then it doesn't need to be  member.

File: vm-virtio/src/console.rs [hunk 5]
```diff
@@ -372,6 +372,7 @@ impl VirtioConsoleConfig {
 
 /// Virtio device for exposing console to the guest OS through virtio.
 pub struct Console {
+    id: String,
```
Comment (rbradford) [inline vm-virtio/src/console.rs:375]: Ditto.

File: vm-virtio/src/console.rs [hunk 6]
```diff
@@ -618,30 +621,26 @@ impl VirtioDevice for Console {
 }
 
 virtio_pausable!(Console);
-const CONSOLE_SNAPSHOT_ID: &str = "virtio-console";
 impl Snapshottable for Console {
     fn id(&self) -> String {
-        CONSOLE_SNAPSHOT_ID.to_string()
+        self.id.clone()
```
Comment (rbradford) [inline vm-virtio/src/console.rs:626]: You could do `String::new("_console0")`

File: vm-virtio/src/console.rs [hunk 7]
```diff
@@ -618,30 +621,26 @@ impl VirtioDevice for Console {
...
 
     fn snapshot(&self) -> std::result::Result<Snapshot, MigratableError> {
         let snapshot =
             serde_json::to_vec(&self.state()).map_err(|e| MigratableError::Snapshot(e.into()))?;
 
-        let mut console_snapshot = Snapshot::new(CONSOLE_SNAPSHOT_ID);
+        let mut console_snapshot = Snapshot::new(self.id.as_str());
         console_snapshot.add_data_section(SnapshotDataSection {
-            id: format!("{}-section", CONSOLE_SNAPSHOT_ID),
+            id: format!("{}-section", self.id),
```
Comment (rbradford) [inline vm-virtio/src/console.rs:635]: And then do `self.id()`

File: vm-virtio/src/console.rs (hunk 5 above)
Comment (sboeuf) [reply vm-virtio/src/console.rs:375]: Oh no this is different here, I want to keep it this way because the virtio console device might be reused by some other VMMs, and they wouldn't want the console id to be hardcoded.
It's on purpose that I set the identifiers from the DeviceManager because it controls the IDs and it can ensure they're unique.

File: vm-virtio/src/console.rs (hunk 6 above)
Comment (sboeuf) [reply vm-virtio/src/console.rs:626]: I really don't want the id to be chosen by the device.

File: vmm/src/device_manager.rs [hunk 8]
```diff
@@ -1686,7 +1686,7 @@ impl DeviceManager {
 
         Ok((
             Arc::clone(&vsock_device) as VirtioDeviceArc,
-            false,
+            vsock_cfg.iommu,
```
Comment (rbradford) [inline vmm/src/device_manager.rs:1725]: The commit message for this commit is kinda wrong :-)

File: vmm/src/device_manager.rs [hunk 9]
```diff
@@ -1415,14 +1416,16 @@ impl DeviceManager {
...
+
             let virtio_rng_device = Arc::new(Mutex::new(
                 vm_virtio::Rng::new(rng_path, rng_config.iommu)
                     .map_err(DeviceManagerError::CreateVirtioRng)?,
             ));
             devices.push((
                 Arc::clone(&virtio_rng_device) as VirtioDeviceArc,
-                false,
-                None,
+                rng_config.iommu,
```
Comment (rbradford) [inline vmm/src/device_manager.rs:1451]: Split this into two. One part to fix the iommu bug and then another to assign the name. For the commit message on the latter don't refer to "--rng" as that's confusing as that it a CLI command and this is independent of the CLI.

File: vmm/src/device_manager.rs [hunk 10]
```diff
@@ -1710,7 +1711,8 @@ impl DeviceManager {
     ) -> DeviceManagerResult<Vec<(VirtioDeviceArc, bool, Option<String>)>> {
         let mut devices = Vec::new();
 
-        let mm = &self.memory_manager.lock().unwrap();
+        let mm = self.memory_manager.clone();
+        let mm = mm.lock().unwrap();
         if let (Some(region), Some(resize)) = (&mm.virtiomem_region, &mm.virtiomem_resize) {
```
Comment (rbradford) [inline vmm/src/device_manager.rs:1751]: I don't see why this needed to be changed?

File: vmm/src/device_manager.rs [hunk 11]
```diff
@@ -75,6 +75,7 @@ const VFIO_DEVICE_NAME_PREFIX: &str = "vfio";
 
 const DISK_DEVICE_NAME_PREFIX: &str = "disk";
 const FS_DEVICE_NAME_PREFIX: &str = "fs";
+const MEM_DEVICE_NAME_PREFIX: &str = "mem";
```
Comment (rbradford) [inline vmm/src/device_manager.rs:81]: This isn't really a prefix but a fixed name.

File: vmm/src/device_manager.rs [hunk 12]
```diff
@@ -77,6 +77,7 @@ const DISK_DEVICE_NAME_PREFIX: &str = "disk";
 const FS_DEVICE_NAME_PREFIX: &str = "fs";
 const NET_DEVICE_NAME_PREFIX: &str = "net";
 const PMEM_DEVICE_NAME_PREFIX: &str = "pmem";
+const RNG_DEVICE_NAME_PREFIX: &str = "rng";
```
Comment (rbradford) [inline vmm/src/device_manager.rs:84]: This isn't a prefix but a fixed name.

File: vmm/src/device_manager.rs [hunk 13]
```diff
@@ -73,6 +73,7 @@ const MMIO_LEN: u64 = 0x1000;
 #[cfg(feature = "pci_support")]
 const VFIO_DEVICE_NAME_PREFIX: &str = "vfio";
 
+const CONSOLE_DEVICE_NAME_PREFIX: &str = "console";
```
Comment (rbradford) [inline vmm/src/device_manager.rs:78]: Ditto not a prefix but a fixed name.

File: vmm/src/device_manager.rs [hunk 14]
```diff
@@ -1099,8 +1100,8 @@ impl DeviceManager {
                     .map_err(DeviceManagerError::CreateVirtioConsole)?;
             virtio_devices.push((
                 Arc::new(Mutex::new(virtio_console_device)) as VirtioDeviceArc,
-                false,
-                None,
+                console_config.iommu,
```
Comment (rbradford) [inline vmm/src/device_manager.rs:1120]: Please split this bug fix into a separate commit.

File: vmm/src/device_manager.rs [hunk 15]
```diff
@@ -82,6 +82,9 @@ const PMEM_DEVICE_NAME_PREFIX: &str = "pmem";
 const RNG_DEVICE_NAME_PREFIX: &str = "rng";
 const VSOCK_DEVICE_NAME_PREFIX: &str = "vsock";
 
+#[cfg(feature = "pci_support")]
+const IOMMU_DEVICE_NAME_PREFIX: &str = "iommu";
+
```
Comment (rbradford) [inline vmm/src/device_manager.rs:89]: Ditto.

File: vmm/src/device_manager.rs [hunk 16]
```diff
@@ -73,6 +73,8 @@ const MMIO_LEN: u64 = 0x1000;
 #[cfg(feature = "pci_support")]
 const VFIO_DEVICE_NAME_PREFIX: &str = "vfio";
 
+const IOAPIC_DEVICE_NAME_PREFIX: &str = "ioapic";
```
Comment (rbradford) [inline vmm/src/device_manager.rs:75]: Ditto.

File: vmm/src/device_manager.rs (hunk 10 above)
Comment (sboeuf) [reply vmm/src/device_manager.rs:1751]: I moved it to another commit but this is needed to avoid the following failure:
```
error[E0502]: cannot borrow `*self` as mutable because it is also borrowed as immutable                                                                                                                                                                              
    --> vmm/src/device_manager.rs:1761:13                                                                                                                                                                                                                            
     |                                                                                                                                                                                                                                                               
1740 |         let mm = self.memory_manager.lock().unwrap();                                                                                                                                                                                                         
     |                  ------------------- immutable borrow occurs here                                                                                                                                                                                             
...                                                                                                                                                                                                                                                                  
1761 |             self.add_migratable_device(Arc::clone(&virtio_mem_device) as Arc<Mutex<dyn Migratable>>);                                                                                                                                                         
     |             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ mutable borrow occurs here                                                                                                                               
...                                                                                                                                                                                                                                                                  
1765 |     }                                                     
     |     - immutable borrow might be used here, when `mm` is dropped and runs the `Drop` code for type `std::sync::MutexGuard`

error: aborting due to previous error

```

**Comment by sboeuf** [comment]
@rbradford I addressed all your comments :)

File: vmm/src/device_manager.rs (hunk 9 above)
Comment (rbradford) [reply vmm/src/device_manager.rs:1451]: Can you do the commit message fix too?

File: vmm/src/device_manager.rs [hunk 17]
```diff
@@ -1100,7 +1101,7 @@ impl DeviceManager {
             virtio_devices.push((
                 Arc::new(Mutex::new(virtio_console_device)) as VirtioDeviceArc,
```
Comment (rbradford) [inline vmm/src/device_manager.rs]: Ditto about the commit message on this one.

**Comment by sboeuf** [comment]
@rbradford oh I just realized I forgot to make the names "internal" by adding the `_`. I'm gonna do that now.
Also, I'll update the commit messages :)

**Comment by sboeuf** [comment]
@rbradford ok I'm done with this PR :)
